pkg/reconciler/nscert/controller.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nscert
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	nsinformer "knative.dev/pkg/injection/informers/kubeinformers/corev1/namespace"
+	"knative.dev/serving/pkg/apis/networking"
+	kcertinformer "knative.dev/serving/pkg/client/injection/informers/networking/v1alpha1/certificate"
+	routecfg "knative.dev/serving/pkg/reconciler/route/config"
+
+	"knative.dev/serving/pkg/network"
+	pkgreconciler "knative.dev/serving/pkg/reconciler"
+	"knative.dev/serving/pkg/reconciler/nscert/config"
+)
+
+const (
+	controllerAgentName = "namespace-controller"
+)
+
+type configStore interface {
+	ToContext(ctx context.Context) context.Context
+	WatchConfigs(w configmap.Watcher)
+}
+
+// NewController initializes the controller and is called by the generated code
+// Registers eventhandlers to enqueue events.
+func NewController(ctx context.Context, cmw configmap.Watcher) *controller.Impl {
+	nsInformer := nsinformer.Get(ctx)
+	knCertificateInformer := kcertinformer.Get(ctx)
+
+	c := &reconciler{
+		Base:                pkgreconciler.NewBase(ctx, controllerAgentName, cmw),
+		nsLister:            nsInformer.Lister(),
+		knCertificateLister: knCertificateInformer.Lister(),
+	}
+
+	impl := controller.NewImpl(c, c.Logger, "Namespace")
+
+	c.Logger.Info("Setting up event handlers")
+	nsInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: pkgreconciler.LabelFilterFunc(networking.WildcardCertLabelKey, "false", true),
+		Handler:    controller.HandleAll(impl.Enqueue),
+	})
+
+	knCertificateInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: controller.Filter(corev1.SchemeGroupVersion.WithKind("Namespace")),
```

Comment:
can we chain filters with `ChainFilters` here to filter the ones with `networking.WildcardCertLabelKey` and then we can remove that check in the reconcile.

Reply:
It's not clear to me how to use `ChainFilters` to express "accept certificates with an owner of type Namespace AND whose owner DOES NOT have a label present". 

The check in reconcile would still need to be there since `GlobalResync` is called on config updates.

Reply:
`ChainFilters` just chains predicates, nothing stops you from writing the one that will check for any condition you want :)

Reply:
I'll do that. I think that would make things clearer.